File: client-scalaz/src/main/scala/org/openapitools/client/api/Application.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime

import java.time.OffsetDateTime

import Application._

case class Application (
  id: Option[Long],
active: Option[Boolean],
valid: Option[Boolean],
title: Option[String],
about: Option[String],
certificateUrl: Option[String],
certificateName: Option[String],
certificate: Option[BlobFile],
certificateLastUpdated: Option[OffsetDateTime],
appDescription: Option[String],
showInActivities: Option[Boolean],
cssClass: Option[String],
inviteWelcomeText: Option[String],
bundleId: Option[String],
appType: Option[String],
scoringType: Option[ScoringType],
appKey: Option[String],
appRestKey: Option[String],
appTokenKey: Option[String],
hintCost: Option[Integer],
eulaDate: Option[OffsetDateTime],
eulaVersion: Option[String],
landingPageUrl: Option[String],
invitePageUrl: Option[String],
downloadUrls: Option[String],
appName: Option[String],
appIcon: Option[Asset],
appLogo: Option[Asset],
billableEntity: Option[BillableEntity],
facebookAppId: Option[String],
facebookAppSecret: Option[String],
twitterAppId: Option[String],
twitterAppSecret: Option[String],
twitterCallbackURL: Option[String],
defaultImage: Option[String],
modules: Option[String],
authorizedServers: Option[String],
authorizedCount: Option[Integer],
activeServers: Option[String],
licensePeriod: Option[OffsetDateTime],
hasGameData: Option[Boolean],
currentBuildVersion: Option[String],
currentApiVersion: Option[String],
globalTickets: Option[Boolean],
ticketsPerPoint: Option[Float],
maxPointsPerScore: Option[Integer],
useAppCache: Option[Boolean],
uniqueUserTag: Option[String],
useMatchingAlgorithm: Option[Boolean],
platforms: Option[List[Platform]],
categories: Option[List[Category]],
placements: Option[List[Placement]],
hasAdvertisements: Option[Boolean],
appleAppId: Option[String],
googleAppId: Option[String],
googleApiKey: Option[String],
googleOAuthClientId: Option[String],
googleOAuthSecret: Option[String],
googleOAuthAccessToken: Option[String],
googleOAuthRefreshToken: Option[String],
baiduApiKey: Option[String],
baiduSecretKey: Option[String],
jpushAppKey: Option[String],
jpushMasterSecret: Option[String],
xgPushSecretKey: Option[String],
publicNotifications: Option[Boolean],
urlScheme: Option[String],
retailer: Option[Retailer],
objectStoreSession: Option[String],
publicContentApproval: Option[Boolean],
productionMode: Option[Boolean],
griplyBaseUrl: Option[String],
minimumSessionLength: Option[Integer],
sessionGapLength: Option[Integer],
processSessionData: Option[Boolean],
localAdsEnabled: Option[Boolean],
includeInReport: Option[Boolean],
sqootApiKey: Option[String],
amqpEnabled: Option[Boolean],
trilatSettings: Option[TrilatAppSettings])

object Application {
  import DateTimeCodecs._
  sealed trait ScoringType
  case object GAMELEVEL extends ScoringType
  case object GAMEOBJECT extends ScoringType

  object ScoringType {
    def toScoringType(s: String): Option[ScoringType] = s match {
      case "GAMELEVEL" => Some(GAMELEVEL)
      case "GAMEOBJECT" => Some(GAMEOBJECT)
      case _ => None
    }

    def fromScoringType(x: ScoringType): String = x match {
      case GAMELEVEL => "GAMELEVEL"
      case GAMEOBJECT => "GAMEOBJECT"
    }
  }

  implicit val ScoringTypeEnumEncoder: EncodeJson[ScoringType] =
    EncodeJson[ScoringType](is => StringEncodeJson(ScoringType.fromScoringType(is)))

  implicit val ScoringTypeEnumDecoder: DecodeJson[ScoringType] =
    DecodeJson.optionDecoder[ScoringType](n => n.string.flatMap(jStr => ScoringType.toScoringType(jStr)), "ScoringType failed to de-serialize")

  implicit val ApplicationCodecJson: CodecJson[Application] = CodecJson.derive[Application]
  implicit val ApplicationDecoder: EntityDecoder[Application] = jsonOf[Application]
  implicit val ApplicationEncoder: EntityEncoder[Application] = jsonEncoderOf[Application]
}

File: client-clojure/src/sirqul_io_t_platform/specs/application.clj
(ns sirqul-io-t-platform.specs.application
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [sirqul-io-t-platform.specs.blob-file :refer :all]
            [sirqul-io-t-platform.specs.asset :refer :all]
            [sirqul-io-t-platform.specs.asset :refer :all]
            [sirqul-io-t-platform.specs.billable-entity :refer :all]
            [sirqul-io-t-platform.specs.platform :refer :all]
            [sirqul-io-t-platform.specs.category :refer :all]
            [sirqul-io-t-platform.specs.placement :refer :all]
            [sirqul-io-t-platform.specs.retailer :refer :all]
            [sirqul-io-t-platform.specs.trilat-app-settings :refer :all]
            )
  (:import (java.io File)))


(def application-data
  {
   (ds/opt :id) int?
   (ds/opt :active) boolean?
   (ds/opt :valid) boolean?
   (ds/opt :title) string?
   (ds/opt :about) string?
   (ds/opt :certificateUrl) string?
   (ds/opt :certificateName) string?
   (ds/opt :certificate) blob-file-spec
   (ds/opt :certificateLastUpdated) inst?
   (ds/opt :appDescription) string?
   (ds/opt :showInActivities) boolean?
   (ds/opt :cssClass) string?
   (ds/opt :inviteWelcomeText) string?
   (ds/opt :bundleId) string?
   (ds/opt :appType) string?
   (ds/opt :scoringType) string?
   (ds/opt :appKey) string?
   (ds/opt :appRestKey) string?
   (ds/opt :appTokenKey) string?
   (ds/opt :hintCost) int?
   (ds/opt :eulaDate) inst?
   (ds/opt :eulaVersion) string?
   (ds/opt :landingPageUrl) string?
   (ds/opt :invitePageUrl) string?
   (ds/opt :downloadUrls) string?
   (ds/opt :appName) string?
   (ds/opt :appIcon) asset-spec
   (ds/opt :appLogo) asset-spec
   (ds/opt :billableEntity) billable-entity-spec
   (ds/opt :facebookAppId) string?
   (ds/opt :facebookAppSecret) string?
   (ds/opt :twitterAppId) string?
   (ds/opt :twitterAppSecret) string?
   (ds/opt :twitterCallbackURL) string?
   (ds/opt :defaultImage) string?
   (ds/opt :modules) string?
   (ds/opt :authorizedServers) string?
   (ds/opt :authorizedCount) int?
   (ds/opt :activeServers) string?
   (ds/opt :licensePeriod) inst?
   (ds/opt :hasGameData) boolean?
   (ds/opt :currentBuildVersion) string?
   (ds/opt :currentApiVersion) string?
   (ds/opt :globalTickets) boolean?
   (ds/opt :ticketsPerPoint) float?
   (ds/opt :maxPointsPerScore) int?
   (ds/opt :useAppCache) boolean?
   (ds/opt :uniqueUserTag) string?
   (ds/opt :useMatchingAlgorithm) boolean?
   (ds/opt :platforms) (s/coll-of platform-spec)
   (ds/opt :categories) (s/coll-of category-spec)
   (ds/opt :placements) (s/coll-of placement-spec)
   (ds/opt :hasAdvertisements) boolean?
   (ds/opt :appleAppId) string?
   (ds/opt :googleAppId) string?
   (ds/opt :googleApiKey) string?
   (ds/opt :googleOAuthClientId) string?
   (ds/opt :googleOAuthSecret) string?
   (ds/opt :googleOAuthAccessToken) string?
   (ds/opt :googleOAuthRefreshToken) string?
   (ds/opt :baiduApiKey) string?
   (ds/opt :baiduSecretKey) string?
   (ds/opt :jpushAppKey) string?
   (ds/opt :jpushMasterSecret) string?
   (ds/opt :xgPushSecretKey) string?
   (ds/opt :publicNotifications) boolean?
   (ds/opt :urlScheme) string?
   (ds/opt :retailer) retailer-spec
   (ds/opt :objectStoreSession) string?
   (ds/opt :publicContentApproval) boolean?
   (ds/opt :productionMode) boolean?
   (ds/opt :griplyBaseUrl) string?
   (ds/opt :minimumSessionLength) int?
   (ds/opt :sessionGapLength) int?
   (ds/opt :processSessionData) boolean?
   (ds/opt :localAdsEnabled) boolean?
   (ds/opt :includeInReport) boolean?
   (ds/opt :sqootApiKey) string?
   (ds/opt :amqpEnabled) boolean?
   (ds/opt :trilatSettings) trilat-app-settings-spec
   })

(def application-spec
  (ds/spec
    {:name ::application
     :spec application-data}))
